examples/platform/silabs/SiWx917/SiWx917/sl_wifi_if.cpp

```diff
@@ -632,115 +621,82 @@ static void wfx_rsi_save_ap_info() // translation
-        ap.security      = connect_security_mode;
-        ap.encryption    = SL_WIFI_NO_ENCRYPTION;
-        ap.credential_id = id;
-        if ((status = sl_wifi_connect(SL_WIFI_CLIENT_INTERFACE, &ap, timeout_ms)) == SL_STATUS_IN_PROGRESS)
-        {
-            callback_status = SL_STATUS_IN_PROGRESS;
-            while (callback_status == SL_STATUS_IN_PROGRESS)
-            {
-                osThreadYield();
-            }
-            status = callback_status;
-        }
-        else
-        {
-            if (is_wifi_disconnection_event || wfx_rsi.join_retries <= WFX_RSI_CONFIG_MAX_JOIN)
-            {
-                SILABS_LOG("wfx_rsi_do_join: Wifi connect failed with status: %x", status);
-                SILABS_LOG("wfx_rsi_do_join: starting JOIN to %s after %d tries\n", (char *) &wfx_rsi.sec.ssid[0],
-                           wfx_rsi.join_retries);
-                wfx_rsi.join_retries += 1;
-                wfx_rsi.dev_state &= ~(WFX_RSI_ST_STA_CONNECTING | WFX_RSI_ST_STA_CONNECTED);
-                wfx_retry_interval_handler(is_wifi_disconnection_event, wfx_rsi.join_retries);
-                if (is_wifi_disconnection_event || wfx_rsi.join_retries <= MAX_JOIN_RETRIES_COUNT)
-                {
-                    event.eventType = WFX_EVT_STA_START_JOIN;
-                    WfxPostEvent(&event);
-                }
-            }
-        }
-    }
+    VerifyOrReturnError(strlen(wfx_rsi.sec.passkey) > SL_WIFI_MAX_PSK_LENGTH, SL_STATUS_SI91X_INVALID_PSK_LENGTH);
+    sl_net_credential_id_t id = SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID;
+    status                    = sl_net_set_credential(id, SL_NET_WIFI_PSK, &wfx_rsi.sec.passkey[0], strlen(wfx_rsi.sec.passkey));
+    VerifyOrReturnError(status == SL_STATUS_OK, status);
+
+    uint32_t timeout_ms = 0;
+
+    ap.ssid.length = strlen(wfx_rsi.sec.ssid);
+    memcpy(ap.ssid.value, (int8_t *) &wfx_rsi.sec.ssid[0], ap.ssid.length);
+    ap.encryption    = SL_WIFI_NO_ENCRYPTION;
+    ap.credential_id = id;
+    status           = sl_wifi_connect(SL_WIFI_CLIENT_INTERFACE, &ap, timeout_ms);
+    // sl_wifi_connect returns SL_STATUS_IN_PROGRESS if join is in progress
+    // after the initial scan is done, the scan does not check for SSID
+    ReturnErrorCodeIf((status == SL_STATUS_OK || status == SL_STATUS_IN_PROGRESS), status);
+
+    // failure only happens when the firmware returns an error
+    ChipLogError(DeviceLayer, "wfx_rsi_do_join: sl_wifi_connect failed: 0x%lx", status);
+    VerifyOrReturnError((is_wifi_disconnection_event || wfx_rsi.join_retries <= MAX_JOIN_RETRIES_COUNT), status);
+
+    wfx_rsi.dev_state &= ~(WFX_RSI_ST_STA_CONNECTING | WFX_RSI_ST_STA_CONNECTED);
+    ChipLogProgress(DeviceLayer, "wfx_rsi_do_join: retry attempt %d", wfx_rsi.join_retries);
+    wfx_retry_interval_handler(is_wifi_disconnection_event, wfx_rsi.join_retries);
+    event.eventType = WFX_EVT_STA_START_JOIN;
+    WfxPostEvent(&event);
+    wfx_rsi.join_retries++;
```

Comment:
Suggest posting at the very end, as posting could wake a different thread, which would see `wfx_rsi.join_retries` non-incremented (if it ever cared). In general, always post as the last thing of a method.